src/main_macro.rs

```diff
@@ -29,23 +31,68 @@
 ///     println!("{}", x);
 /// });
 /// ```
+///
+/// With command line arguments.
+///
+/// ```rust,ignore
+/// #[macro_use] extern crate quicli;
+/// use quicli::prelude::*;
+/// use std::path::PathBuf;
+///
+/// #[derive(Debug, StructOpt)]
+/// struct Cli {
+///     /// The config file for the cli app.
+///     #[structopt(long = "config", short = "c", parse(from_os_str))]
+///     config: PathBuf,
+/// }
+///
+/// main!(|args: Cli| {
+///     let x = read_file(&args.config)?;
+///     println!("{}", x);
+/// });
+/// ```
+///
+/// With command line arguments including a verbosity flag.
+///
+/// ```rust,ignore
+/// #[macro_use] extern crate quicli;
+/// use quicli::prelude::*;
+/// use std::path::PathBuf;
+///
+/// #[derive(Debug, StructOpt)]
+/// struct Cli {
+///     /// The config file for the cli app.
+///     #[structopt(long = "config", short = "c", parse(from_os_str))]
+///     config: PathBuf,
+///     #[structopt(flatten)]
+///     verbosity: Verbosity,
+/// }
+///
+/// main!(|args: Cli, log_level: Verbosity| {
+///     let x = read_file(&args.config)?;
+///     println!("{}", x);
+/// });
+/// ```
 #[macro_export]
 macro_rules! main {
     (|$args:ident: $cli:ty, log_level: $verbosity:ident| $body:expr) => {
         fn main() {
-            fn run() -> $crate::prelude::Result<()> {
-                let $args = <$cli>::from_args();
-                $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME"))?;
-
+            #[allow(unused_variables)]
+            fn run($args: $cli) -> $crate::prelude::Result<()> {
                 $body
-
+                #[allow(unreachable_code)]
                 Ok(())
             }
 
-            if let Err(e) = run() {
-                eprintln!("error: {}", e);
+            let $args = <$cli>::from_args();
+            // This cannot fail because we control the main method, so no-one else can register a
+            // log handler before us.
```

Comment:
Is that the only reason log handlers can fail?

Reply:
Yes - I looked through the methods and the only failure point is `log::set_boxed_logger` (called by `env_logger`), and that can only fail if a logger has already been set.